### PR TITLE
GSPO-token policy loss function

### PR DESCRIPTION
**Pull request by nkkarpov**
## Description

GSPO-token policy loss function from https://arxiv.org/pdf/2507.18071

## Checklist

Please check the following items before code is ready to be reviewed.

- [x]  Code has passed all tests
- [x]  Docstrings have been added/updated in Google Style
- [x]  Documentation has been updated
- [x]  Code is ready for review

### CODE REVIEW

File: trinity/algorithm/policy_loss_fn/gspo_policy_loss.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        self,
+        logprob: torch.Tensor,
+        old_logprob: torch.Tensor,
+        action_mask: torch.Tensor,
+        advantages: torch.Tensor,
+        **kwargs,
+    ) -> Tuple[torch.Tensor, Dict]:
+        negative_approx_kl = logprob - old_logprob
+        seq_lengths = torch.sum(action_mask, dim=-1).clamp(min=1).unsqueeze(-1)
+        negative_approx_kl = negative_approx_kl / seq_lengths
```
Comment (hiyuchang) [inline trinity/algorithm/policy_loss_fn/gspo_policy_loss.py]:
```suggestion
        negative_approx_kl_seq = torch.sum(negative_approx_kl * action_mask, dim=-1) / seq_lengths
```

File: trinity/algorithm/policy_loss_fn/gspo_policy_loss.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        log_seq_importance_ratio = logprob - logprob.detach() + negative_approx_kl.detach()
```
Comment (hiyuchang) [inline trinity/algorithm/policy_loss_fn/gspo_policy_loss.py]:
```suggestion
        log_seq_importance_ratio = logprob - logprob.detach() + negative_approx_kl_seq.detach().unsqueeze(-1)
```

File: trinity/algorithm/policy_loss_fn/gspo_policy_loss.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        ratio = torch.exp(log_seq_importance_ratio)
+        ppo_kl = masked_mean(-negative_approx_kl, action_mask)
+
+        pg_losses = -advantages * ratio
+        pg_losses2 = -advantages * torch.clamp(
+            ratio, 1.0 - self.clip_range_low, 1.0 + self.clip_range_high
+        )
+
+        pg_loss = masked_mean(torch.max(pg_losses, pg_losses2), action_mask)
```
Comment (hiyuchang) [inline trinity/algorithm/policy_loss_fn/gspo_policy_loss.py]:
```suggestion
        seq_losses = torch.sum(torch.max(pg_losses, pg_losses2) * action_mask, dim=-1) / torch.sum(action_mask, dim=-1)
        pg_loss = torch.mean(seq_losses)
```

Comment (nkkarpov) [reply trinity/algorithm/policy_loss_fn/gspo_policy_loss.py]: I agree with renaming, but I don't see the benefit of removing ```masked_mean``` call.  In general, it is good to reuse the code, especially if it adds some numerical stability (I assume).

File: trinity/algorithm/policy_loss_fn/gspo_policy_loss.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+            "pg_clipfrac": pg_clipfrac.detach().item(),
+            "ppo_kl": ppo_kl.detach().item(),
+            "pg_loss": pg_loss.detach().item(),
+        }
+        return pg_loss, metrics
+
+    @classmethod
+    def default_args(cls) -> Dict:
+        return {
+            "clip_range": 0.2,
```
Comment (yanxi-chen) [inline trinity/algorithm/policy_loss_fn/gspo_policy_loss.py]:
```suggestion
            "clip_range_low": 0.0003,
            "clip_range_high": 0.0004,
```

File: trinity/algorithm/policy_loss_fn/gspo_policy_loss.py (hunk 3 above)
Comment (hiyuchang) [reply trinity/algorithm/policy_loss_fn/gspo_policy_loss.py]: Yes, you are right. It's better to use:

```suggestion
        seq_losses = masked_mean(torch.max(pg_losses, pg_losses2), action_mask, axis=-1)
        pg_loss = torch.mean(seq_losses)
```

**Comment by hiyuchang** [comment]
Thanks for your contribution! We have suggested some revisions, with a reference to [verl's PR about GSPO](https://github.com/volcengine/verl/pull/2775/files). 

Could you verify if the suggested modifications are correct, update the unittest if needed, and then rerun the pre-commit and unittest?

File: trinity/algorithm/policy_loss_fn/gspo_policy_loss.py (hunk 4 above)
Comment (nkkarpov) [reply trinity/algorithm/policy_loss_fn/gspo_policy_loss.py]: I have tried to reuse constants from the original ```ppo_policy_loss```. Do you have any justification beyond these values?

Comment (yanxi-chen) [reply trinity/algorithm/policy_loss_fn/gspo_policy_loss.py]: These new values are adopted from Section 5.1 of the GSPO paper, which are indeed quite different from the conventional value 0.2 for PPO/GRPO. I suppose it's mainly due to the difference between clipping prob ratio at token vs sequence level. Some related discussions can be found [here](https://github.com/volcengine/verl/pull/2775#issuecomment-3130065984) too.

**Comment by nkkarpov** [comment]
Thank you, @hiyuchang and @yanxi-chen , for the feedback. I've made another pass and improved the naming and verbosity of the code. I also tried to balance creating new variables and keeping lines concise.

File: trinity/algorithm/policy_loss_fn/gspo_policy_loss.py (hunk 4 above)
Comment (nkkarpov) [reply trinity/algorithm/policy_loss_fn/gspo_policy_loss.py]: Thanks for the comment, I have added the comment with link to the discussion!

File: trinity/algorithm/policy_loss_fn/gspo_policy_loss.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        self,
+        logprob: torch.Tensor,
+        old_logprob: torch.Tensor,
+        action_mask: torch.Tensor,
+        advantages: torch.Tensor,
+        **kwargs,
+    ) -> Tuple[torch.Tensor, Dict]:
+        seq_lengths = action_mask.sum(dim=-1, keepdim=True).clamp_min(1)
+        negative_approx_kl_seq = (logprob - old_logprob) * action_mask / seq_lengths
+        log_seq_importance_ratio = logprob - logprob.detach() + negative_approx_kl_seq.detach()
```
Comment (yanxi-chen) [inline trinity/algorithm/policy_loss_fn/gspo_policy_loss.py]:
```suggestion
        negative_approx_kl = logprob - old_logprob
        negative_approx_kl_seq = torch.sum(negative_approx_kl * action_mask, dim=-1) / seq_lengths
        log_seq_importance_ratio = logprob - logprob.detach() + negative_approx_kl_seq.detach().unsqueeze(-1)
```

File: trinity/algorithm/policy_loss_fn/gspo_policy_loss.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        
+
+        pg_losses = -advantages * ratio
+        pg_losses_clipped = -advantages * torch.clamp(
+            ratio, 1.0 - self.clip_range_low, 1.0 + self.clip_range_high
+        )
+        
+        pg_loss = masked_mean(torch.max(pg_losses, pg_losses_clipped), action_mask)
+        pg_clipfrac = masked_mean(torch.gt(pg_losses_clipped, pg_losses).float(), action_mask)
+        ppo_kl = masked_mean(-negative_approx_kl_seq, action_mask)
```
Comment (yanxi-chen) [inline trinity/algorithm/policy_loss_fn/gspo_policy_loss.py]:
```suggestion
        ppo_kl = masked_mean(-negative_approx_kl, action_mask)
```

File: trinity/algorithm/policy_loss_fn/gspo_policy_loss.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        log_seq_importance_ratio = logprob - logprob.detach() + negative_approx_kl_seq.detach()
+        ratio = torch.exp(log_seq_importance_ratio)
+        
+
+        pg_losses = -advantages * ratio
+        pg_losses_clipped = -advantages * torch.clamp(
+            ratio, 1.0 - self.clip_range_low, 1.0 + self.clip_range_high
+        )
+        
+        pg_loss = masked_mean(torch.max(pg_losses, pg_losses_clipped), action_mask)
```
Comment (yanxi-chen) [inline trinity/algorithm/policy_loss_fn/gspo_policy_loss.py]:
```suggestion
        seq_losses = masked_mean(torch.max(pg_losses, pg_losses_clipped), action_mask, axis=-1)
        pg_loss = torch.mean(seq_losses)
```

File: trinity/algorithm/policy_loss_fn/gspo_policy_loss.py (hunk 5 above)
Comment (yanxi-chen) [reply trinity/algorithm/policy_loss_fn/gspo_policy_loss.py]: This modification is due to the mechanism of clipping sequence-wise prob ratio (rather than token-wise) in GSPO. The sequence-wise log prob difference (which equals sum of token-wise log prob difference) is calculated by `torch.sum(negative_approx_kl * action_mask, dim=-1)`.

File: trinity/algorithm/policy_loss_fn/gspo_policy_loss.py (hunk 7 above)
Comment (yanxi-chen) [reply trinity/algorithm/policy_loss_fn/gspo_policy_loss.py]: This modification follows the formulation in the GSPO paper, i.e., the loss is an average over losses of sequences, and the loss of each sequence is an average over its token-wise losses.

**Review by yanxi-chen — COMMENTED** [review]
Please see whether the inline comments better reflect the algorithm logic as described in the GSPO paper, and update the unittest if necessary. Otherwise lgtm!

**Comment by nkkarpov** [comment]
@yanxi-chen My purpose was to implement GSPO-token loss function. Please, find the attached image from the original paper. Additionallym I have tried to make the interface compatible with the original grpo policy loss function (it accepts the same tensors). I assume you looked into `vanilla` GSPO loss function, and it probably lead you to the confusion. 
Screenshot 2025-08-01 at 12 29 19

**Comment by yanxi-chen** [comment]
> @yanxi-chen My purpose was to implement GSPO-token loss function. Please, find the attached image from the original paper. Additionallym I have tried to make the interface compatible with the original grpo policy loss function (it accepts the same tensors). I assume you looked into `vanilla` GSPO loss function, and it probably lead you to the confusion. Screenshot 2025-08-01 at 12 29 19

@nkkarpov Thanks for the update. 

I understand that you intend to implement the GSPO-token variant. Note that Eq. (14) still involves $s_i(\theta)$ which, as defined in Eq. (7), involves **sequence-wise prob ratio** (which equals product of token-wise prob ratio). 
image

You can also see this in the gradient of the GSPO-token loss, i.e., Eq. (16) and (17) of the paper:
image

This explains my proposed modification: `negative_approx_kl_seq = torch.sum(negative_approx_kl * action_mask, dim=-1) / seq_lengths`. Please feel free to correct me if I'm wrong :)

**Comment by nkkarpov** [comment]
Yes, you are right, now it should be correct :)

**Comment by yanxi-chen** [comment]
Thanks @nkkarpov . Could you fix the pre-commit errors and then commit again? Otherwise I think it's ready to merge :)

**Review by yanxi-chen — APPROVED** [review]
lgtm, thanks for contributing!
